### PR TITLE
[metricbuilder] hide details of pdata

### LINKED ISSUE #8358: [Metrics builder] Generate whole pdata.Metrics object
`NewMetricData` method was added in https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/8255 but we can avoid exposing it and create pdata.Metrics internally by the first call to Record...DataPoint. Emit signature should be changed to:

```go
func (mb *MetricsBuilder) Emit() pdata.Metrics
```

Originally posted: https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/8255#discussion_r821014408

### PR DESCRIPTION
**Pull request by codeboten**
Refactoring the metric builder to hide details of pdata. This added a couple of methods to the metric builder interface but i think that's ok.

Fix https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/8358

@dmitryax @djaglowski please take a look as you're familiar with the original issue.

### CODE REVIEW

File: receiver/mysqlreceiver/testdata/scraper/expected.json [hunk 1]
```diff
@@ -4,7 +4,7 @@
          "instrumentationLibraryMetrics": [
             {
                "instrumentationLibrary": {
-                  "name": "otel/mysql"
+                  "name": "otelcol/mysqlreceiver"
```
Comment (dmitryax) [inline receiver/mysqlreceiver/testdata/scraper/expected.json:7]: Shouldn't this be a part of another PR?

File: cmd/mdatagen/metrics_v2.tmpl [hunk 2]
```diff
@@ -133,17 +134,40 @@ func NewMetricsBuilder(settings MetricsSettings, options ...metricBuilderOption)
...
+	}
 	{{- range $name, $metric := .Metrics }}
-	mb.metric{{- $name.Render }}.emit(metrics)
+	mb.metric{{- $name.Render }}.emit(mb.data.ResourceMetrics().At(0).InstrumentationLibraryMetrics().At(0).Metrics())
 	{{- end }}
+	return *mb.data
+}
+
+// EnsureCapacity TODO
+func (mb *MetricsBuilder) EnsureCapacity(length int) {
```
Comment (dmitryax) [inline cmd/mdatagen/metrics_v2.tmpl:148]: I don't think we need these functions. Capacity can be always taken from number of enabled metrics

File: cmd/mdatagen/metrics_v2.tmpl [hunk 3]
```diff
@@ -101,6 +101,7 @@ func newMetric{{ $name.Render }}(settings MetricSettings) metric{{ $name.Render
 // MetricsBuilder provides an interface for scrapers to report metrics while taking care of all the transformations
 // required to produce metric representation defined in metadata and user settings.
 type MetricsBuilder struct {
+	data                     *pdata.Metrics  // data buffer for generated metric.
```
Comment (dmitryax) [inline cmd/mdatagen/metrics_v2.tmpl:104]: Each field `metric{{ $name.Render }}` is already a buffer. So I don't think we need this. We can create `pdata.Metrics` in the `Emit` call (sorry the the confusing description in the issue that I added)

File: cmd/mdatagen/metrics_v2.tmpl [hunk 4]
```diff
@@ -133,17 +134,40 @@ func NewMetricsBuilder(settings MetricsSettings, options ...metricBuilderOption)
...
 }
 
 {{ range $name, $metric := .Metrics -}}
 // Record{{ $name.Render }}DataPoint adds a data point to {{ $name }} metric.
 func (mb *MetricsBuilder) Record{{ $name.Render }}DataPoint(ts pdata.Timestamp
 	{{- if $metric.Data.HasMetricValueType }}, val {{ $metric.Data.MetricValueType.BasicType }}{{ end }}
 	{{- range $metric.Attributes -}} , {{ .RenderUnexported }}AttributeValue string{{ end }}) {
+	if mb.data == nil {
+		mb.data = mb.newMetricData()
+	}
```
Comment (dmitryax) [inline cmd/mdatagen/metrics_v2.tmpl:170]: This is probably not needed

File: cmd/mdatagen/metrics_v2.tmpl [hunk 5]
```diff
@@ -133,17 +134,40 @@ func NewMetricsBuilder(settings MetricsSettings, options ...metricBuilderOption)
...
 // another set of data points. This function will be doing all transformations required to produce metric representation
 // defined in metadata and user settings, e.g. delta/cumulative translation.
-func (mb *MetricsBuilder) Emit(metrics pdata.MetricSlice) {
+func (mb *MetricsBuilder) Emit() pdata.Metrics{
+	if mb.data == nil {
+		return pdata.NewMetrics()
+	}
 	{{- range $name, $metric := .Metrics }}
-	mb.metric{{- $name.Render }}.emit(metrics)
+	mb.metric{{- $name.Render }}.emit(mb.data.ResourceMetrics().At(0).InstrumentationLibraryMetrics().At(0).Metrics())
```
Comment (dmitryax) [inline cmd/mdatagen/metrics_v2.tmpl:142]: Can we create a var `metrics := mb.data.ResourceMetrics().At(0).InstrumentationLibraryMetrics().At(0).Metrics()` and use it as an argument to the `emit` calls to improve readability of the generated code?

File: receiver/mysqlreceiver/testdata/scraper/expected.json (hunk 1 above)
Comment (codeboten) [reply receiver/mysqlreceiver/testdata/scraper/expected.json:7]: If i change the Emit interface I end up having to change the receiver that weren't already refactored to use `NewMetricData`. I can migrate the receivers (mysql, zookeeper) to NewMetricData as a middle step to this change.

Comment (dmitryax) [reply receiver/mysqlreceiver/testdata/scraper/expected.json:7]: > I can migrate the receivers (mysql, zookeeper) to NewMetricData as a middle step to this change.

I though that was the plan. But it's up to you

Comment (dmitryax) [reply receiver/mysqlreceiver/testdata/scraper/expected.json:7]: I vote for splitting it just to remove "breaking changes" from this PR to keep it cleaner.

File: cmd/mdatagen/metrics_v2.tmpl (hunk 2 above)
Comment (dmitryax) [reply cmd/mdatagen/metrics_v2.tmpl:148]: Or we can follow similar approach to how we figure out capacity for datapoints from the first scrape.

**Review by djaglowski — COMMENTED** [review]
This change looks really nice from the perspective of a scraper author. There's little to no "overhead" left within the ScrapeFunc.

File: receiver/mysqlreceiver/testdata/scraper/expected.json (hunk 1 above)
Comment (codeboten) [reply receiver/mysqlreceiver/testdata/scraper/expected.json:7]: I split them into #8387 and #8389 👍

Comment (codeboten) [reply receiver/mysqlreceiver/testdata/scraper/expected.json]: Rebased.

File: cmd/mdatagen/metrics_v2.tmpl (hunk 4 above)
Comment (codeboten) [reply cmd/mdatagen/metrics_v2.tmpl]: Removed

File: cmd/mdatagen/metrics_v2.tmpl (hunk 3 above)
Comment (codeboten) [reply cmd/mdatagen/metrics_v2.tmpl]: Please take a look at the update implementation @dmitryax, I followed your suggestions and think it's much cleaner

File: receiver/hostmetricsreceiver/internal/scraper/processscraper/process_scraper.go [hunk 6]
```diff
@@ -83,27 +82,24 @@ func (s *scraper) start(context.Context, component.Host) error {
...
-	for _, md := range metadata {
-		rm := rms.AppendEmpty()
-		rm.SetSchemaUrl(conventions.SchemaURL)
-		md.initializeResource(rm.Resource())
-		metrics := rm.InstrumentationLibraryMetrics().AppendEmpty().Metrics()
+	resource := pdata.NewResource()
+
+	for _, md := range mdata {
+
+		md.initializeResource(resource)
```
[the remark this answers is not in the document]
Comment (djaglowski) [reply receiver/hostmetricsreceiver/internal/scraper/processscraper/process_scraper.go:102]: One solution would be to create another builder at the resource level, and move the Record... functions to that. 

The scraper author can ask the MetricsBuilder for a new ResourceBuilder as needed, attach attributes and data points to it, and then ultimately call `Emit` on the MetricsBuilder. Since the MetricsBuilder created the ResourceBuilders, it can call `emit` on each of those in turn.

[the remark this answers is not in the document]
Comment (codeboten) [reply receiver/hostmetricsreceiver/internal/scraper/processscraper/process_scraper.go:102]: `Emit` can be called with an option `WithResource` which copies the resource into the generate metrics, would that not work here?

https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/8382/files#diff-806ca51937857b079bdde5bf7e3e87aeb773245b2cb93b40078d241515bdab68R138-R142

[the remark this answers is not in the document]
Comment (codeboten) [reply receiver/hostmetricsreceiver/internal/scraper/processscraper/process_scraper.go:102]: Ah i see.. Yes the code is limited to a single resource for each call to Emit which means the only way to support multiple resources at this time is with separate calls to emit. Let me see what this looks like with a resource builder

[the remark this answers is not in the document]
Comment (dmitryax) [reply receiver/hostmetricsreceiver/internal/scraper/processscraper/process_scraper.go:102]: I'm thinking a bit different approach to this problem that also solves https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/7301. Let me submit an alternative PR and we can continue the discussion then

**Comment by codeboten** [comment]
superseeded by https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/8555
